pkg/ingester/errors.go

```diff
@@ -77,114 +60,255 @@ func (e errorWithStatus) GRPCStatus() *status.Status {
 	return e.status
 }
 
+// safeToWrapError is a marker interface for the errors that are safe to wrap.
+type safeToWrapError interface {
+	safeToWrap()
+}
+
+// badDataError is a marker interface for the errors representing bad data.
+type badDataError interface {
+	badData()
+}
+
 // wrapOrAnnotateWithUser prepends the given userID to the given error.
-// If the error is safe, the returned error retains a reference to the former.
+// If the given error matches one of the errors from this package, the
+// returned error retains a reference to the former.
 func wrapOrAnnotateWithUser(err error, userID string) error {
-	// If this is a safe error, we wrap it with userID and return it, because
-	// it might contain extra information for gRPC and our logging middleware.
-	var safe safeToWrap
-	if errors.As(err, &safe) {
+	// If err is a safeToWrapError, we wrap it with userID and return it.
+	var safeToWrap safeToWrapError
+	if errors.As(err, &safeToWrap) {
 		return fmt.Errorf("user=%s: %w", userID, err)
 	}
+
 	// Otherwise, we just annotate it with userID and return it.
 	return fmt.Errorf("user=%s: %s", userID, err)
 }
 
-func newIngestErrSample(errID globalerror.ID, errMsg string, timestamp model.Time, labels []mimirpb.LabelAdapter) error {
-	return safeToWrapError(
-		fmt.Sprintf("%v. The affected sample has timestamp %s and is from series %s",
-			errID.Message(errMsg),
+// validationError is an error indicating a problem with a sample or an exemplar.
+type validationError struct {
+	message string
+}
+
+func (e validationError) Error() string {
+	return e.message
+}
+
+func (e validationError) safeToWrap() {}
+
+func (e validationError) badData() {}
+
+func newSampleError(prefixMsg string, timestamp model.Time, labels []mimirpb.LabelAdapter) validationError {
+	return validationError{
+		message: fmt.Sprintf(
+			"%s. The affected sample has timestamp %s and is from series %s",
+			prefixMsg,
 			timestamp.Time().UTC().Format(time.RFC3339Nano),
 			mimirpb.FromLabelAdaptersToLabels(labels).String(),
 		),
-	)
+	}
```

Comment:
Thank you, I have re-organized the code accordingly.

Reply:
This is not quite complete. There's still unnecessary string formatting done when calling `globalerror.ID.Message(...)` too early. This should also be delayed to `Error` calls.

Reply:
I wouldn't touch error sampling now. This is a separate problem, and it would additionally complicate this PR and this task, which is basically to remove `httpgrpc.Errorf()` calls from ingester and distributor.

Reply:
>This is not quite complete. There's still unnecessary string formatting done when calling globalerror.ID.Message(...) too early. This should also be delayed to Error calls.

I was asked by @colega to do it this way [here](https://github.com/grafana/mimir/pull/6324#discussion_r1355069739). If your opinion about formatting is strong, I will put the globalID and message back to the error as fields, and format them during `Error()`.

Reply:
Yes, I would suggest to put it back. While I see Oleg's point, creating additional strings in "new" call is premature, and can be avoided by putting formatting into `Error` call. If `Error` call is actually never called, then we just saved formatting completely.

Reply:
I will fix this immediately.